python/ray/dashboard/client/src/pages/dashboard/node-info/features/GRAM.tsx

```diff
@@ -0,0 +1,74 @@
+import React from "react";
+import { GPUStats } from "../../../../api";
+import { MiBRatio } from "../../../../common/formatUtils";
+import UsageBar from "../../../../common/UsageBar";
+import { getWeightedAverage } from "../../../../common/util";
+import {
+  ClusterFeatureComponent,
+  Node,
+  NodeFeatureComponent,
+  WorkerFeatureComponent,
+} from "./types";
+
+const nodeGRAMUtilization = (node: Node) => {
+  const utilization = (gpu: GPUStats) => gpu.memory_used / gpu.memory_total;
+  const utilizationSum = node.gpus.reduce(
+    (acc, gpu) => acc + utilization(gpu),
+    0,
+  );
+  const avgUtilization = utilizationSum / node.gpus.length;
+  // Convert to a percent before returning
+  return avgUtilization * 100;
+};
+
+const clusterGRAMUtilization = (nodes: Array<Node>) => {
+  return getWeightedAverage(
+    nodes.map((node) => ({
+      weight: node.gpus.length,
+      value: nodeGRAMUtilization(node),
+    })),
+  );
+};
+
+export const ClusterGRAM: ClusterFeatureComponent = ({ nodes }) => {
+  const clusterAverageUtilization = clusterGRAMUtilization(nodes);
+  return (
+    <div style={{ minWidth: 60 }}>
+      <UsageBar
+        percent={clusterAverageUtilization}
+        text={`${clusterAverageUtilization.toFixed(1)}%`}
+      />
+    </div>
+  );
+};
+
+export const NodeGRAM: NodeFeatureComponent = ({ node }) => {
+  const gramUtil = nodeGRAMUtilization(node);
+  return (
+    <div style={{ minWidth: 60 }}>
+      <UsageBar percent={gramUtil} text={`${gramUtil.toFixed(1)}%`} />
+    </div>
+  );
+};
+
+export const WorkerGRAM: WorkerFeatureComponent = ({ worker, node }) => {
+  const workerProcessPerGPU = node.gpus
+    .map((gpu) => gpu.processes)
+    .map((processes) =>
+      processes.find((process) => process.pid === worker.pid),
+    );
+  const workerUtilPerGPU = workerProcessPerGPU.map(
+    (proc) => proc?.gpu_memory_usage || 0,
+  );
+  const totalGRAMperGPU = node.gpus.map((gpu) => gpu.memory_total);
+  const totalNodeGRAM = totalGRAMperGPU.reduce((acc, usage) => acc + usage, 0);
```

Comment:
What does `acc` stand for? Can we use a different variable name for this?

Reply:
Acc stands for accumulator and is the traditional name for the value that gets built up during a call to reduce as I've seen, but I'm definitely open to suggestions. Maybe I could call it `sum`?

Reply:
I'm going to replace all these usages with a sum function that I define in `util.ts`. I wouldn't normally write sum using reduce, but ts didn't have a built in sum function.